internal/postgres/data.go

```diff
@@ -0,0 +1,266 @@
+// based on the specified Spanner and PostgreSQL schemas. Note that since entries
+// in vals may be empty, we also return the list of columns (empty cols are dropped).
+func ConvertData(conv *Conv, spTable, pgTable string, cols []string, vals []string) ([]string, []interface{}, error) {
+	spannerSchema, ok1 := conv.spSchema[spTable]
+	pgSchema, ok2 := conv.pgSchema[pgTable]
+	if !ok1 || !ok2 {
+		return []string{}, []interface{}{}, fmt.Errorf("Can't find table %s in schema", spTable)
+	}
+	var c []string
+	var v []interface{}
+	if len(cols) != len(vals) {
+		return []string{}, []interface{}{}, fmt.Errorf("Bad parameters: cols and vals have different lengths: len(cols)=%d, len(vals)=%d", len(cols), len(vals))
+	}
+	for i, col := range cols {
+		if vals[i] == "\\N" { // PostgreSQL representation of empty column in COPY-FROM blocks.
+			continue
+		}
+		cd, ok := spannerSchema.Cds[col]
+		if !ok {
+			return []string{}, []interface{}{}, fmt.Errorf("Can't find col %s in schema", col)
+		}
+		pgCol, ok := pgSchema.cols[col]
+		if !ok {
+			return []string{}, []interface{}{}, fmt.Errorf("Can't find col %s in pgSchema", col)
+		}
+		var x interface{}
+		var err error
+		if cd.IsArray {
+			x, err = convArray(cd.T, pgCol, conv.location, vals[i])
+		} else {
+			x, err = convScalar(cd.T, pgCol.id, conv.location, vals[i])
+		}
+		if err != nil {
+			return []string{}, []interface{}{}, err
+		}
+		v = append(v, x)
+		c = append(c, col)
+	}
+	if aux, ok := conv.syntheticPKeys[spTable]; ok {
+		c = append(c, aux.col)
+		v = append(v, int64(bits.Reverse64(uint64(aux.sequence))))
+		aux.sequence++
+		conv.syntheticPKeys[spTable] = aux
+	}
+	return c, v, nil
+}
+
+func convScalar(spannerType ddl.ScalarType, pgType string, location *time.Location, val string) (interface{}, error) {
+	// Whitespace within the val string is considered part of the data value.
+	// Note that many of the underlying conversions functions we use (like
+	// strconv.ParseFloat and strconv.ParseInt) return "invalid syntax"
+	// errors if whitespace were to appear at the start or end of a string.
+	// We do not expect pg_dump to generate such output.
+	switch spannerType.(type) {
+	case ddl.Bool:
+		return convBool(val)
+	case ddl.Bytes:
+		return convBytes(val)
+	case ddl.Date:
+		return convDate(val)
+	case ddl.Float64:
+		return convFloat64(val)
+	case ddl.Int64:
+		return convInt64(val)
+	case ddl.String:
+		return val, nil
+	case ddl.Timestamp:
+		return convTimestamp(pgType, location, val)
+	default:
+		return val, fmt.Errorf("Data conversion not implemented for type %v", reflect.TypeOf(spannerType))
+	}
+}
+
+func convBool(val string) (bool, error) {
+	b, err := strconv.ParseBool(val)
+	if err != nil {
+		return b, fmt.Errorf("Can't convert to bool: %w", err)
+	}
+	return b, err
+}
+
+func convBytes(val string) ([]byte, error) {
+	if val[0:3] != `\\x` {
+		return []byte{}, fmt.Errorf("Can't convert bytea data to bytes: doesn't start with \\x prefix")
```

Comment:
typo? 
s/bytea data to/data to/

Reply:
Done.

Actually this was intended: 'bytea' is the PostgreSQL type we are converting from here (https://www.postgresql.org/docs/9.0/datatype-binary.html).
But after looking at the other comments in the convXXX functions, I think it makes sense to have all of these errors use a consistent format.